server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -0,0 +1,164 @@
+        RoutingAllocation allocation,
+        Logger logger
+    ) {
+        HashMap<ShardRouting, AllocateUnassignedDecision> shardAllocationDecisions = new HashMap<>();
+        final boolean explain = allocation.debugDecision();
+        Set<ShardRouting> shardsEligibleForFetch = new HashSet<>();
+        Set<ShardRouting> shardsNotEligibleForFetch = new HashSet<>();
+        // identify ineligible shards
+        for (ShardRouting shard : shards) {
+            AllocateUnassignedDecision decision = getInEligibleShardDecision(shard, allocation);
+            if (decision != null) {
+                shardsNotEligibleForFetch.add(shard);
+                shardAllocationDecisions.put(shard, decision);
+            } else {
+                shardsEligibleForFetch.add(shard);
+            }
+        }
+        // Do not call fetchData if there are no eligible shards
+        if (shardsEligibleForFetch.size() == 0) {
+            return shardAllocationDecisions;
+        }
+        // only fetch data for eligible shards
+        final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(
+            shardsEligibleForFetch,
+            shardsNotEligibleForFetch,
+            allocation
+        );
+
+        // process the received data
+        for (ShardRouting unassignedShard : shardsEligibleForFetch) {
+            if (shardsState.hasData() == false) {
+                // if fetching is not done, add that no decision in the resultant map
+                allocation.setHasPendingAsyncFetch();
+                List<NodeAllocationResult> nodeDecisions = null;
+                if (explain) {
+                    nodeDecisions = buildDecisionsForAllNodes(unassignedShard, allocation);
+                }
+                shardAllocationDecisions.put(
+                    unassignedShard,
+                    AllocateUnassignedDecision.no(AllocationStatus.FETCHING_SHARD_DATA, nodeDecisions)
+                );
+            } else {
+
+                NodeShardStates nodeShardStates = adaptToNodeShardStates(unassignedShard, shardsState);
+                // get allocation decision for this shard
+                shardAllocationDecisions.put(unassignedShard, getAllocationDecision(unassignedShard, allocation, nodeShardStates, logger));
+            }
+        }
+        return shardAllocationDecisions;
+    }
+
+    /**
+     * shardsState contain the Data, there key is DiscoveryNode but value is Map<ShardId,
+     * NodeGatewayStartedShardsBatch> so to get one shard level data (from all the nodes), we'll traverse the map
+     * and construct the nodeShardState along the way before making any allocation decision. As metadata for a
+     * particular shard is needed from all the discovery nodes.
+     *
+     * @param unassignedShard unassigned shard
+     * @param shardsState fetch data result for the whole batch
+     * @return shard state returned from each node
+     */
+    private static NodeShardStates adaptToNodeShardStates(
```

Comment:
Trying to understand, are we traversing `shardsState` Map of Map for each `unassignedShard` ? Can this exercise not be done in a single pass, building the exhaustive view for allocation decisions for all shards across all Discovery Nodes ?

Reply:
Decision making requires data of a single shard only. Right now we're doing for loop on nodes then get with shardId in `adaptToNodeShardStates` method.
So even if we pre-compute the data of all the shards, the overall compute will be same as it is right now. (i.e. right now we've an outer loop on batch of shards, and inner loop on nodes. Later we'll have an outer loop of nodes and inner loop of shards)

Currently `adaptToNodeShardStates` looks a bit more readable as it constructs the data for a single shard and return it. So right now structure is also a bit simple just `List<NodeShardState>`. If we pre-compute, it'll be changed to `Map<ShardRouting, List<NodeShardState>>`. 

Considering the readability and same compute, currently keeping it as it is.